### PR TITLE
[optim] skip .item calls in all optimizers when compiling with dynamo

### LINKED ISSUE #1083: Remove .item() calls from optimizer.step()
There are a lot of ```.item()``` calls inside of ```optimizer.step()```, e.g:
https://github.com/pytorch/pytorch/blob/master/torch/optim/asgd.py#L182-L212

These ```.item()``` calls inside of the for loop cause graph break when enabling dynamo and finally hitting cache limit. I took a closer look at these usages and found actually we can remove these call because the results were feed into other downstream tensor operations.
I did a simple verification on ```ASGD``` and found removing ```.item()``` works well and no graph break on Dynamo end. 

I'd like to check if there are other constraints to prevent us from doing this. If not, I can go ahead to fix this.

cc @jansel

### PR DESCRIPTION
**Pull request by soumith**
@mlazos: skips `item()` calls if compiling with dynamo, by defining a helper function `_get_value` which either returns the result of `.item()` or the scalar cpu tensor if compiling with dynamo. This was done because removing `item()` calls significantly regresses eager perf. Additionally, `_dispatch_sqrt` calls the appropriate sqrt function (math.sqrt, or torch.sqrt).

Fixes https://github.com/pytorch/torchdynamo/issues/1083

This PR will no longer be needed once symint support is default. 

This PR closes all remaining graph breaks in the optimizers (!!)

cc @mlazos @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by soumith** [comment]
after this PR lands, I suspect we can remove the `capturable` keyword and what it represents.

**Review by albanD — COMMENTED** [review]
Does this make it capturable? These are still CPU Tensors.

**Comment by soumith** [comment]
> Also this will trigger https://github.com/pytorch/pytorch/issues/74424 again. So not sure if we can do this.

I dont think it will trigger that issue with XLA because we've removed all `item` calls. We can even get rid of most list comprehensions. So, I have high hopes that this wont be an issue for XLA

**Comment by soumith** [comment]
> Does this make it capturable? These are still CPU Tensors.

Capturable (i believe) is whether dynamo can get a whole graph of it -- not that it should all be on GPU.

**Comment by albanD** [comment]
> Capturable (i believe) is whether dynamo can get a whole graph of it

Capturable is used in the context of cudagraph: "whether this instance is safe to capture in a CUDA graph.".
So we do need the step to be on the GPU when this is true.

**Comment by albanD** [comment]
I think there

**Comment by soumith** [comment]
> Capturable is used in the context of cudagraph: "whether this instance is safe to capture in a CUDA graph.".

Ho, interesting. In that case, I have to check `capturable` code, it's probably buggy as-is because when `learning_rate` drops happen, the scalar changes / drops but cudagraph would have baked it in

**Comment by albanD** [comment]
I think there are two things here:

1) This patch
Scalar Tensor on CPU will be slower than plain python number. I guess we're happy with a small hit here but I think it would be good to measure it as some of these optimizer do quite a few ops (and our framework overhead is non-trivial).

2) The capturable flag:
I think there is still an impasse here though:
- For this optimizer to be capturable by cudagraph, we MUST have the step on the GPU
- To get ok performance outside of cudagraph, we MUST keep the step on the CPU

We could consider checking during the step itself if we're using cudagraph but that would mean that we might have to move the step then (which is a sync point so not good).

**Comment by mlazos** [comment]
@albanD One question: if the step is on GPU, this would be bad without cudagraphs because we're launching a kernel to do step + 1 essentially right? I think in the dynamo case this won't matter, because this op will get fused into a larger kernel anyway so could we make `capturable=True` by default, and then for people who aren't using dynamo or cudagraphs, they can manually set `capturable=False`?

**Comment by albanD** [comment]
> this would be bad without cudagraphs because we're launching a kernel to do step + 1 essentially right?

Correct.

> could we make capturable=True by default, and then for people who aren't using dynamo or cudagraphs, they can manually set capturable=False?

We can't make it the default no. For GPU-like backends, having these small  ops on the device is prohibitively slow https://github.com/pytorch/pytorch/issues/74424

File: torch/optim/adamax.py
Comment (albanD) [inline torch/optim/adamax.py:297]: torch.stack

File: torch/optim/adamw.py
Comment (albanD) [inline torch/optim/adamw.py:419]: torch.stack

File: torch/optim/asgd.py
Comment (albanD) [inline torch/optim/asgd.py:263]: Isn't the original code plain wrong?? All the params are updates with a single eta from the 0th value.

File: torch/optim/nadam.py
Comment (albanD) [inline torch/optim/nadam.py:264]: TODO: we can remove most (if not all) of the if statement here now that this side doesn't use `.item()` either.

File: torch/optim/nadam.py
Comment (albanD) [inline torch/optim/nadam.py:315]: torch.stack

File: torch/optim/asgd.py
Comment (albanD) [inline torch/optim/asgd.py:278]: torch.stack

File: torch/optim/radam.py
Comment (albanD) [inline torch/optim/radam.py:294]: torch.stack

**Review by albanD — COMMENTED** [review]
Sounds good to me!

**Comment by byronyi** [comment]
> For GPU-like backends, having these small ops on the device is prohibitively slow #74424

I guess XLA is not GPU-like backends as it is lazy?

> in each training iteration, now state['step'] needs to be move back to CPU from CUDA or XLA to compute things like step = step_t.item()

Seems the overhead mainly coming from **moving those tensors back to CPU**, which is exactly the issue this PR fixes.

**Comment by JackCaoG** [comment]
Generally speaking removing `item` call will be only beneficial to the xla backend. `item` call is one of the common cause of early execution. Without it, lazy will fuse all optimizer + forward + backward. I am not entirely sure what `capturable` does, @shunting314 might be able to comment for our `dynamo->xla` bridge.

**Comment by shunting314** [comment]
> @shunting314 might be able to comment for our `dynamo->xla` bridge.

in the dynamoc -> xla bridge for training, we are not running optimizer actually ( https://github.com/pytorch/pytorch/blob/master/benchmarks/dynamo/common.py#L894 ). But looks like people are adding them.

**Comment by byronyi** [comment]
> I am not entirely sure what `capturable` does

If `capturable` is True then the step is a device tensor instead of put on CPU. See https://github.com/pytorch/pytorch/blob/master/torch/optim/adam.py#L212-L216

File: torch/optim/asgd.py
Comment (mlazos) [reply torch/optim/asgd.py:300]: So I think the reason this works is if all of the etas are on the same device, etas[:] will all hold the same value. eta is only a funciton lambd, lr, step, and alpha which are constants per group

Comment (mlazos) [reply torch/optim/asgd.py:300]: I'm also fine with fixing it though

File: torch/optim/nadam.py
Comment (mlazos) [reply torch/optim/nadam.py:265]: Dynamo doesn't like the `value` argument since the scalars aren't  traceable, is it okay if I go with the then clause of the `if`? Similar question with the addcdiv_ in Adam. I can also post numbers showing that Adam single tensor with dynamo + inductor is faster than the multitensor version if this helps with justification

File: torch/optim/adam.py
Comment (mlazos) [reply torch/optim/adam.py:507]: `torch.tensor` is indeed still shady

File: torch/optim/adam.py
Comment (mlazos) [inline torch/optim/adam.py:420]: Converting step to a tensor induces the last remaining test failure, any idea why that might be @albanD ? I played around with `torch.pow` and it looks like there is a difference between scalar exponentiation and torch exponentiation that differs from the fused Adam impl. (Maybe fused Adam matched the scalar pow?)

Comment (soumith) [reply torch/optim/adam.py:420]: scalar exponentiation is done in `double` precision, whereas default tensorization is float32. exponentiation is particularly sensitive to the reduced precision. I'd verify this by moving `step` to `float64`

Comment (mlazos) [reply torch/optim/adam.py:420]: Thanks! I will try this

File: torch/optim/nadam.py
Comment (mlazos) [reply torch/optim/nadam.py]: Resolved offline, I have this PR here: https://github.com/pytorch/pytorch/pull/90227 should fix this.

File: torch/optim/adam.py
Comment (mlazos) [reply torch/optim/adam.py:420]: Hmm this didn't work, the accuracy actually went down (I got NaNs). Originally 22/64 elements are mismatches, and now they all are.

Comment (mlazos) [reply torch/optim/adam.py:420]: Ignore my previous comment, I confirmed this by casting to double at the exponentiation (I just cast step at line 422) and it worked. Should I try and make step float64 by default, or increase the tolerance of the test

File: test/test_cuda.py
Comment (albanD) [inline test/test_cuda.py:2456]: What is this function?

File: test/test_cuda.py
Comment (albanD) [inline test/test_cuda.py:2453]: Why does this need to be updated? Would passing `exact_dtype=False` to `assertEqual` be enough?

File: torch/optim/adam.py
Comment (albanD) [inline torch/optim/adam.py:200]: Note that this won't change older saved states on optimizers. It might be simpler to keep this the same and change the dtype within the step?

File: torch/optim/asgd.py
Comment (albanD) [reply torch/optim/asgd.py:302]: We should fix it for sure. But this PR is not the right place for this as it need proper testing/etc

File: torch/optim/nadam.py
Comment (albanD) [inline torch/optim/nadam.py:255]: This comment is out of date.

File: torch/optim/nadam.py
Comment (albanD) [inline torch/optim/nadam.py:260]: value=1.0 is the default, we can remove it from here.

File: torch/optim/radam.py
Comment (albanD) [inline torch/optim/radam.py:355]: torch.stack?

**Comment by albanD** [comment]
Do we have the sanity check benchmark as discussed above to make sure nothing is obviously broken in terms of speed due to these changes?

File: test/test_cuda.py
Comment (mlazos) [reply test/test_cuda.py:2456]: Yeah I just looked up the proper way of casting

File: test/test_cuda.py
Comment (mlazos) [reply test/test_cuda.py]: Actually removed this once I put the cast into `step`

**Comment by mlazos** [comment]
Update: The current version matches the eager perf of the old version. Should be good to go, just need a signoff. [Updated nums](https://gist.github.com/mlazos/e84cc17e080db5c5ebae8446c0072e68)

File: torch/_dynamo/config.py
Comment (albanD) [inline torch/_dynamo/config.py:63]: How is that different from the one above: `external_utils.is_compiling` ?

File: torch/optim/nadam.py
Comment (albanD) [inline torch/optim/nadam.py:266]: Why do you do the `_get_value` on mu_product here and not on `mu_product_next` like in the old code?

File: torch/optim/optimizer.py
Comment (albanD) [inline torch/optim/optimizer.py:40]: Why scripting is involved here?

**Review by albanD — APPROVED** [review]
That is acceptable but really not great in term of readability.

**Comment by albanD** [comment]
By the way @mlazos I guess that https://github.com/pytorch/pytorch/pull/90624 should actually allow you to remove the .item() changes?
What about the other changes? I think that it is part of the dynamo goal to be able to do these integrations without modifying user code with if is_compiling a-la torchscipt?

**Comment by mlazos** [comment]
I don't think symint is ready for prime time since there are still some lingering bugs. Once dynamic shapes is on by default I'm fine with removing this. 

Yes we usually don't want to modify code to compile with dynamo. In most other cases we don't care about eager perf as much though. Like in this case if we didn't care about eager perf I'd just remove the item calls and call it a day since dynamo provides good speed up. It ends up being a trade off off of is this worth modifying the compiler and adding x feature vs should I just remove x from the code if it isn't really necessary. With user code I lean towards the former while in code we own it's more balanced.

File: torch/_dynamo/config.py
Comment (mlazos) [reply torch/_dynamo/config.py:63]: It actually isn't, but I spend like half the day Friday trying to resolve the circular dependency issue with importing dynamo in the optimizer code and couldn't resolve it. It came down to the config that dynamo uses importing a bunch of other torch modules which made it impossible to remove. Other workarounds I tried:

* inlining import (bad for performance)
* loading import once with a helper function (bad for performance)
* just calling torch._dynamo.is_compiling without importing like we do with is_scripting (caused weird import issues)

This ended up being the least intrusive, since this is temporary thought this was the best option.

File: torch/optim/optimizer.py
Comment (mlazos) [reply torch/optim/optimizer.py:40]: Yeah the tests last night were failing. TorchScript expects each variable to only have a single type and each function to only have a single return type, so I gave TorchScript this "type hint" essentially.

**Comment by albanD** [comment]
>  In most other cases we don't care about eager perf as much though.

I would challenge that actually. People that write library code (nn, vision, audio, etc) need to support both. So this is a major use case!

**Pull request closed without merging**
